File: tests/sample-code/vulnerable.py
"""Sample vulnerable Python code for testing Semgrep rules"""

import os
import hashlib
import random
from subprocess import call

# SHOULD TRIGGER: hardcoded-password
DATABASE_CONFIG = {
    'host': 'localhost',
    'user': 'admin',
    'password': 'admin123',
    'database': 'production_db'
}

# SHOULD TRIGGER: exec-command-injection
def execute_system_command(user_input):
    """Dangerous: executes arbitrary commands"""
    os.system(f"ls {user_input}")
    return True

# SHOULD TRIGGER: sql-injection-risk
def get_user(username):
    """Vulnerable to SQL injection"""
    query = "SELECT * FROM users WHERE username = '" + username + "'"
    # execute(query)  # Simulated
    return query

# SHOULD TRIGGER: weak-crypto-md5
def hash_user_password(password):
    """Using weak MD5 hashing"""
    return hashlib.md5(password.encode()).hexdigest()

# SHOULD TRIGGER: insecure-random
def generate_session_token():
    """Using insecure random for security-sensitive operation"""
    return str(random.random())

# SHOULD TRIGGER: no-print-statements
def debug_function(data):
    """Debug prints that shouldn't be in production"""
    print(f"Debug: {data}")
    print("Processing data...")
    return data

# SHOULD TRIGGER: hardcoded-api-key
API_KEY = "sk_test_1234567890abcdefghijklmnopqrstuvwxyz"
SECRET_KEY = "secret_key_9876543210zyxwvutsrqponmlkjihgfedcba"

# SHOULD TRIGGER: todo-comment
def calculate_discount(price, customer_type):
    # TODO: Implement tiered discount system
    # FIXME: This doesn't work for enterprise customers
    if customer_type == 'premium':
        return price * 0.9
    return price

# Safe code (should NOT trigger)
def safe_password_handling():
    """Example of secure password handling"""
    password = os.getenv('DB_PASSWORD')
    return password

def safe_random():
    """Using cryptographically secure random"""
    import secrets
    return secrets.token_hex(16)

def safe_sql_query(username):
    """Using parameterized queries"""
    query = "SELECT * FROM users WHERE username = ?"
    params = (username,)
    return query, params

if __name__ == "__main__":
    # SHOULD TRIGGER: no-print-statements
    print("Running vulnerable code examples")
    debug_function({"test": "data"})

File: tests/sample-code/vulnerable.js
// Sample vulnerable code for testing Semgrep rules

// SHOULD TRIGGER: dangerous-eval-usage
function unsafeEval(userInput) {
  const result = eval(userInput);
  return result;
}

// SHOULD TRIGGER: hardcoded-password
const config = {
  username: "admin",
  password: "SuperSecret123!",
  apiKey: "sk_test_FakeKey1234567890abcdefghijklmnopqrstuvwxyz" // Fake test key for demonstration
};

// SHOULD TRIGGER: unsafe-innerhtml
function updateContent(userContent) {
  document.getElementById('content').innerHTML = userContent;
}

// SHOULD TRIGGER: insecure-random
function generateToken() {
  return Math.random().toString(36).substring(7);
}

// SHOULD TRIGGER: no-console-log
function debugUser(user) {
  console.log("User data:", user);
  console.debug("Debug info:", user.id);
  return user;
}

// SHOULD TRIGGER: empty-catch-block
async function fetchData() {
  try {
    const response = await fetch('/api/data');
    return await response.json();
  } catch (error) {
    // Empty catch - should be flagged
  }
}

// SHOULD TRIGGER: todo-comment
function calculateTotal(items) {
  // TODO: Add tax calculation
  // FIXME: This doesn't handle discounts properly
  return items.reduce((sum, item) => sum + item.price, 0);
}

// SHOULD TRIGGER: weak-crypto-md5
const crypto = require('crypto');
function hashPassword(password) {
  return crypto.createHash('md5').update(password).digest('hex');
}

// SHOULD TRIGGER: magic-numbers
function processPayment(amount) {
  if (amount > 10000) {
    throw new Error('Amount too large');
  }
  return amount * 1.0825; // Tax rate hardcoded
}

module.exports = {
  unsafeEval,
  config,
  updateContent,
  generateToken,
  debugUser,
  fetchData,
  calculateTotal,
  hashPassword,
  processPayment
};

